dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java

```diff
@@ -0,0 +1,150 @@
+        forceSet(TestProperties.CONTAINER_PORT, "0");
+        return ServletDeploymentContext
+                .builder(new PolymorphicPrincipalInjectedResourceConfig())
+                .initParam(
+                  ServletProperties.JAXRS_APPLICATION_CLASS,
+                  PolymorphicPrincipalInjectedResourceConfig.class.getName())
+                .build();
+    }
+
+    public static class PolymorphicPrincipalInjectedResourceConfig extends AbstractAuthResourceConfig {
+
+        public PolymorphicPrincipalInjectedResourceConfig() {
+            super();
+            register(PolymorphicPrincipalEntityResource.class);
+            packages("io.dropwizard.jersey.jackson");
+        }
+
+        @Override protected Class<? extends Principal> getPrincipalClass() {
+            throw new AssertionError("getPrincipalClass must not be invoked");
+        }
+
+        @Override protected ContainerRequestFilter getAuthFilter() {
+            return new ContainerRequestFilter() {
+                @Override public void filter(ContainerRequestContext requestContext) throws IOException {
+                    throw new AssertionError("getAuthFilter result must not be invoked");
+                }
+            };
+        }
+
+        @Override protected AbstractBinder getAuthBinder() {
+            final Set<Class<? extends Principal>> principalClassSet = new HashSet<>();
+            principalClassSet.add(JsonPrincipal.class);
+            principalClassSet.add(NullPrincipal.class);
+            return new PolymorphicAuthValueFactoryProvider.Binder<Principal>(principalClassSet);
+        }
+
+        @Override protected DynamicFeature getAuthDynamicFeature(ContainerRequestFilter authFilter) {
+            final Authenticator<BasicCredentials, JsonPrincipal> jsonAuthenticator = credentials -> {
+                if (credentials.getUsername().equals(JSON_USERNAME)) {
+                    return Optional.<JsonPrincipal>of(new JsonPrincipal(credentials.getUsername()));
+                } else {
+                    return Optional.empty();
+                }
+            };
+
+            final Authenticator<BasicCredentials, NullPrincipal> nullAuthenticator = credentials -> {
+                if (credentials.getUsername().equals(NULL_USERNAME)) {
+                    return Optional.<NullPrincipal>of(new NullPrincipal());
+                } else {
+                    return Optional.empty();
+                }
+            };
+
+            final BasicCredentialAuthFilter jsonAuthFilter = new BasicCredentialAuthFilter.Builder<JsonPrincipal>()
+                .setAuthenticator(jsonAuthenticator)
+                .buildAuthFilter();
+
+            final BasicCredentialAuthFilter nullAuthFilter = new BasicCredentialAuthFilter.Builder<NullPrincipal>()
+                .setAuthenticator(nullAuthenticator)
+                .buildAuthFilter();
+
+            return new PolymorphicAuthDynamicFeature(ImmutableMap.of(
```

Comment:
To avoid a compiler warning about an unchecked call, we should provide type argument `<Principal>` for `PolymorphicAuthDynamicFeature`.